### DISCUSSION TITLE
What makes the Vercel platform not the ideal choice for a REST/GraphQL API?

### DISCUSSION THREAD

**Discussion by rfgamaral** [vercel/vercel#4720] [General]
Hi there,

Ever since I started developing a GraphQL API on top of the Vercel platform that I've been told that although it's possible, it isn't the best fit for it. There's also a similar question posted [here](https://github.com/vercel/vercel/discussions/4418) recently, but it doesn't explain the reasons why.

Just to be clear, I understand that this is not specific to Vercel, but to most serverless platforms. For instance, Fastify has a [documentation page](https://www.fastify.io/docs/latest/Serverless/) for serverless configurations and they also state that "Fastify is not designed to run on serverless environments", but they also don't explain why, they just state that a serverless environment is different from a serverful one.

What I'd like to take away from this discussion are the **technical reasons** why a serverless platform (like Vercel, AWS Lambda, Google Cloud Functions / Run, and probably others) is not the best fit for a REST/GraphQL API. What exactly makes them different from a serverful platform (like Heroku, Digital Ocean, Vultr, and others)?

**Comment by darekrossman** [comment]
We use Vercel for several GraphQL serverless functions without issue. From what I can tell, the only thing that really matters is whether or not you perform all the work you need in your function in a couple seconds. GraphQL gets called out sometimes as your serverless function needs to instantiate the schema and start up a one-off graphql server for every request. This can take time/memory that could _potentially_ lead to costly executions if they exceed say, 10 seconds or something. 

Apollo calls this out their Federated Schema docs pages:

> We recommend against running your gateway in a serverless environment (such as AWS Lambda), because schema composition increases Apollo Server's startup time. If you do run your gateway in a serverless environment, set your function's timeout to at least 10 seconds to prevent unexpected errors.

There may be other technical reasons/caveats, but Ive not been able to find much. Would love to know what you find out.

**Reply by rfgamaral** [reply]
That's one interesting reason and it's exactly the kind of thing I'm looking for. We need that kind of information to be better documented so that one can make well-informed decisions on the best approach to their own unique scenario. I'm wondering what other technical reasons/caveats (as you put it) are there on running REST / GraphQL APIs.

I don't understand why your answer was marked as accepted. It's a very valid answer but it's only one reason from a user of the platform, not a developer. I'd love to have additional input from Vercel developers themselves and I'm unmarking your answer as accepted (for now) in an attempt to draw more attention to the question. Hope you don't mind that.

**Reply by darekrossman** [reply]
Makes total sense. Id also like to see what the official opinions are from the Vercel team.
